### PR TITLE
[runtime] Add Dedup Support to MkBundle

### PR DESCRIPTION
**Pull request by alexanderkyte**

### CODE REVIEW

File: mcs/build/profiles/monotouch_runtime.make [hunk 1]
```diff
@@ -37,5 +37,5 @@ MOBILE_PROFILE = yes
 NO_CONSOLE = yes
 
 PROFILE_DISABLE_BTLS=1
-MONO_FEATURE_APPLETLS=1
+PROFILE_APPLETLS=1
```
Comment (marek-safar) [inline mcs/build/profiles/monotouch_runtime.make]: Please split this change into separate PR

File: mcs/build/tests.make [hunk 2]
```diff
@@ -141,10 +143,49 @@ ifdef TEST_NUNITLITE_APP_CONFIG_RUNTIME
 	sed -i -e "/__INSERT_CUSTOM_APP_CONFIG_RUNTIME__/r $(TEST_NUNITLITE_APP_CONFIG_RUNTIME)" $(NUNITLITE_CONFIG_FILE)
 endif
 
+ifdef PLATFORM_AOT_SUFFIX
+
+DEDUP_DUMMY_CS=$(topdir)/class/lib/$(PROFILE)/DummyInflated.cs
```
Comment (marek-safar) [inline mcs/build/tests.make]: Use meaningful name

File: mcs/build/tests.make [hunk 3]
```diff
@@ -141,10 +143,49 @@ ifdef TEST_NUNITLITE_APP_CONFIG_RUNTIME
...
+DEDUP_DUMMY=$(topdir)/class/lib/$(PROFILE)/DummyInflated.dll
+
+$(DEDUP_DUMMY):
+	echo " // Empty Assembly \n\n" > $(DEDUP_DUMMY_CS)
+	$(CSCOMPILE) -t:library -out:$(DEDUP_DUMMY) $(DEDUP_DUMMY_CS) 
+	rm $(DEDUP_DUMMY_CS)
+
+MKBUNDLE_TEST_BIN = $(TEST_HARNESS).static
+MKBUNDLE_EXE = $(topdir)/class/lib/$(PROFILE)/mkbundle.exe
+TEST_ASSEMBLIES := $(sort $(patsubst .//%,%,$(filter-out %.exe.static %.dll.dll %.exe.dll %bare% %plaincore% %secxml% %Facades% %ilasm%,$(filter %test.dll nunit%.dll nunit%.exe,$(wildcard $(topdir)/class/lib/$(PROFILE)/*)))))
```
Comment (marek-safar) [inline mcs/build/tests.make]: This looks very hacky, can you at least make easier to notice and update

File: mcs/class/System/Test/System.IO.Compression/DeflateStreamTest.cs [hunk 4]
```diff
@@ -46,19 +46,22 @@ private static bool compare_buffers (byte[] first, byte[] second, int length)
 
 		[Test]
 		[ExpectedException (typeof (ArgumentNullException))]
+    [Category ("StaticLinkedAotNotWorking")] // Native MPH loading issues
```
Comment (marek-safar) [inline mcs/class/System/Test/System.IO.Compression/DeflateStreamTest.cs]: Use same indentation

File: mcs/tools/mkbundle/mkbundle.cs [hunk 5]
```diff
@@ -351,6 +374,78 @@ static int Main (string [] args)
 			case "--bundled-header":
 				bundled_header = true;
 				break;
+			case "--in-tree":
```
Comment (marek-safar) [inline mcs/tools/mkbundle/mkbundle.cs]: Update man page as well

File: mcs/build/tests.make [hunk 6]
```diff
@@ -94,8 +94,10 @@ endif
 
 test_assemblies :=
 
+test_lib_output = $(topdir)/class/lib/$(PROFILE)/$(test_lib)
```
Comment (marek-safar) [inline mcs/build/tests.make]: Why do you need this change?

File: mcs/build/profiles/monotouch_watch.make [hunk 7]
```diff
@@ -9,6 +9,5 @@ NO_THREAD_SUSPEND_RESUME=1
 NO_MULTIPLE_APPDOMAINS=1
 NO_PROCESS_START=1
 NO_MONO_SECURITY=1
-MONO_FEATURE_APPLETLS=
```
Comment (akoeplinger) [inline mcs/build/profiles/monotouch_watch.make]: this changes behavior, MONO_FEATURE_APPLETLS wasn't defined before on monotouch_watch but now it is (due to it inheriting PROFILE_APPLETLS=1 from monotouch_runtime.make)

File: mcs/build/tests.make (hunk 3 above)
Comment (akoeplinger) [inline mcs/build/tests.make]: I agree.

Question: how hard do you think would it be instead of compiling everything into the `class/lib/$(PROFILE)` folder to compile/copy all the assets into a separate directory e.g. `class/lib/test` and run everything from there?

This would have the benefit of not polluting the profile output dir, being more similar to how the products use it and also something I'd like to do longer-term anyway for the other profiles (with Jenkins pipeline we could e.g. zip up the directory, put it on another machine and run suites in parallel across machines).

File: mcs/build/tests.make (hunk 2 above)
Comment (alexanderkyte) [inline mcs/build/tests.make]: This is the Dummy dedup module for inflated methods. I can't think of a much better name than DummyInflated and Dedup_Dummy_CS.

File: mcs/build/profiles/monotouch_runtime.make (hunk 1 above)
Comment (alexanderkyte) [inline mcs/build/profiles/monotouch_runtime.make]: https://github.com/mono/mono/pull/5361

File: mcs/tools/mkbundle/mkbundle.cs (hunk 5 above)
Comment (alexanderkyte) [inline mcs/tools/mkbundle/mkbundle.cs]: This is addressed by the latest commit

File: mcs/build/tests.make (hunk 2 above)
Comment (marek-safar) [inline mcs/build/tests.make]: Why is it needed then (does DedupInflatedMethods.cs makes sense to you) ? Is it per library because this will break parallel build ?

Comment (alexanderkyte) [inline mcs/build/tests.make]: The architecture of dedup works by having all of the other (parallel) compilations skip emitting possibly duplicated methods and then by having one compilation (with no data dependency on the other ones) sweep over all of the source files that skipped methods and emit a single copy of each method into that AOT file.

Comment (alexanderkyte) [inline mcs/build/tests.make]: I'll make it DedupInflatedMethods.cs

File: scripts/ci/run-test-testing_aot_full.sh [hunk 8]
```diff
@@ -1,5 +1,10 @@
 #!/bin/bash -e
 
+if test -n "${MONO_STATIC_AOT}";
+then
+${TESTCMD} --label=mkbundle --timeout=25m make -j 4 -w -C mcs/class/corlib -k mkbundle-all-tests
```
[the remark this answers is not in the document]
Comment (alexanderkyte) [reply scripts/ci/run-test-testing_aot_full.sh]: This doesn't only bundle corlib, this bundles everything. We just need to run out of a subdir for the tests.make target changes to be pulled in.

**Comment by akoeplinger** [comment]
@kyte this is breaking tons of tests, why did you merge it? Reverted it for now :)

**Comment by alexanderkyte** [comment]
I saw the approval from Mareks, built locally to check that there weren't breakages, and merged. What broke after the merge?

**Comment by akoeplinger** [comment]
talked with kyte on slack and explained the breakages.
